generators/importanceSampler.cc

```diff
@@ -0,0 +1,423 @@
+		mMinPar += _masses[part];
+
+		std::stringstream nameM;
+		std::stringstream texM;
+		nameM << "m";
+		texM << "m_{";
+		for (size_t p = 1; p <= part+1; ++p) {
+			nameM << p;
+			texM << p;
+		}
+		texM << "}";
+		BCParameter param(nameM.str(), mMinPar, mMax, texM.str());
+		if (part == _nPart-1) {
+			if (mMin == mMax) {
+				// X mass is fixed to one value
+				param.Fix(mMax);
+			} else {
+				param.SetLimits(mMin, mMax);
+			}
+		}
+		AddParameter(param);
+
+		// sample cos(theta), because dOmega = sin(theta) dtheta dphi = dcos(theta) dphi
+		std::stringstream nameTheta;
+		std::stringstream texTheta;
+		nameTheta << "cosTheta" << part+1;
+		texTheta << "cos(#theta_{" << part+1 << "})";
+		AddParameter(nameTheta.str(), -1., 1., texTheta.str());
+
+		std::stringstream namePhi;
+		std::stringstream texPhi;
+		namePhi << "phi" << part+1;
+		texPhi << "#phi_{" << part+1 << "}";
+		AddParameter(namePhi.str(), 0., 2*M_PI, texPhi.str());
+	}
+	for (size_t part = 0; part < _nPart; ++part) {
+		for (size_t qart = 0; qart < part; ++qart) {
+			std::stringstream name;
+			std::stringstream tex;
+			name << "M" << qart+1 << part+1 << 2;
+			tex << "m^{2}_{" << qart+1 << part+1 << "}";
+			AddObservable(name.str(), 0., mMax*mMax, tex.str());
+		}
+	}
+
+	resetFuncInfo();
+	BCAux::SetStyle();
+}
+
+
+double
+rpwa::importanceSampler::LogAPrioriProbability(const std::vector<double>& parameters)
+{
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+	++(_funcCallInfo[LOGAPRIORIPROBABILITY].nmbCalls);
+	TStopwatch timerTot;
+	timerTot.Start();
+
+	rpwa::nBodyPhaseSpaceKinematics nBodyPhaseSpace;
+	if (not initializeNBodyPhaseSpace(nBodyPhaseSpace, parameters, false)) {
+		return -std::numeric_limits<double>::infinity();
+	}
+
+	double phaseSpace = nBodyPhaseSpace.calcWeight() / std::pow(parameters[3*_nPart-6] - _mSum, _nPart-2);
+	if (_massPrior) {
+#ifdef _OPENMP
+#pragma omp critical(massPrior)
+#endif
+		{
+			phaseSpace *= _massPrior->Eval(parameters[3*_nPart-6]);
+		}
+	}
+
+	timerTot.Stop();
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+	_funcCallInfo[LOGAPRIORIPROBABILITY].totalTime += timerTot.RealTime();
+
+	return std::log(phaseSpace);
+}
+
+
+double
+rpwa::importanceSampler::LogLikelihood(const std::vector<double>& parameters)
+{
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+	++(_funcCallInfo[LOGLIKELIHOOD].nmbCalls);
+	TStopwatch timerTot;
+	timerTot.Start();
+
+	if (_phaseSpaceOnly) {
+		return -1;
+	}
+
+	rpwa::nBodyPhaseSpaceKinematics nBodyPhaseSpace;
+	if (not initializeNBodyPhaseSpace(nBodyPhaseSpace, parameters, true)) {
+		printErr << "error while initializing n-body phase space. Aborting..." << std::endl;
+		throw;
+	}
+
+	nBodyPhaseSpace.calcBreakupMomenta();
+	nBodyPhaseSpace.calcEventKinematics(TLorentzVector(0., 0., 0., parameters[3*_nPart-6]));
+	std::vector<TVector3> decayKinMomenta(_nPart);
+	for (size_t part = 0; part < _nPart; ++part) {
+		decayKinMomenta[part] = nBodyPhaseSpace.daughter(part).Vect();
```

Comment:
there is a `daughters` accessor to get the whole vector, but using this here should not make any difference.